### PR TITLE
Implement failure recovery

### PR DESCRIPTION
**Pull request by karlenDimla**
Task/Issue URL: https://app.asana.com/0/0/1206063664298512/f 

### Description
See attached description

### Steps to test this PR
https://app.asana.com/0/0/1206483769461553/f

### CODE REVIEW

**Comment by karlenDimla** [comment]
* `develop`
  * **#4139** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/4139?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @karlenDimla and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/configuration/WgServerApi.kt [hunk 1]
```diff
@@ -35,7 +38,15 @@ interface WgServerApi {
...
 
-    suspend fun registerPublicKey(publicKey: String): WgServerData?
+    suspend fun registerPublicKey(
+        publicKey: String,
+        mode: Mode = Default,
+    ): WgServerData?
+
+    sealed class Mode {
+        data object Default : Mode()
+        data class FailureRecovery(val currentServer: String) : Mode()
```
Comment (aitorvs) [inline network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/configuration/WgServerApi.kt:48]: Why not overload the `toString` method (like below) so that we don't need to do the mapping between the model and the `MODE_FAILURE` constant and we just do `toString()` on it?

```suggestion
        data class FailureRecovery(val currentServer: String) : Mode() {
            override fun toString(): String {
                return "failureRecovery"
            }
        }
```

File: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/configuration/WgServerApi.kt [hunk 2]
```diff
@@ -117,6 +137,10 @@ class RealWgServerApi @Inject constructor(
         val city: String? by attributes
         val country: String? by attributes
     }
+
+    companion object {
+        private const val MODE_FAILURE = "failureRecovery"
+    }
```
Comment (aitorvs) [inline network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/configuration/WgServerApi.kt:143]: Ditto above, I think we can remove

File: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/configuration/WgServerApi.kt [hunk 3]
```diff
@@ -35,7 +38,15 @@ interface WgServerApi {
         val gateway: String,
     )
 
-    suspend fun registerPublicKey(publicKey: String): WgServerData?
+    suspend fun registerPublicKey(
+        publicKey: String,
+        mode: Mode = Default,
```
Comment (aitorvs) [inline network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/configuration/WgServerApi.kt:43]: I think `mode` should be nullable rather than forcing a value on it and then have a `Default` mode that is (and will never) be used.
```suggestion
        mode: Mode? = null,
```

File: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/configuration/WgServerApi.kt [hunk 4]
```diff
@@ -61,11 +75,17 @@ class RealWgServerApi @Inject constructor(
             }
 
         val userPreferredLocation = netNetpEgressServersProvider.updateServerLocationsAndReturnPreferred()
-        val registerKeyBody = if (selectedServer != null) {
+        val registerKeyBody = if (mode is FailureRecovery) {
+            RegisterKeyBody(publicKey = publicKey, server = mode.currentServer, mode = MODE_FAILURE)
+        } else if (selectedServer != null) {
```
Comment (aitorvs) [inline network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/configuration/WgServerApi.kt:80]: Ditto, I thin we should make `mode` nullable and edit this logic accordingly

File: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/pixels/WireguardHandshakeMonitor.kt
Comment (aitorvs) [inline network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/pixels/WireguardHandshakeMonitor.kt:1]: I don't think the failure recovery logic should be inside the handshake monitor. The monitor was very simple, now it's kind of not.

I'd rather create a `WireguardHandshakeMonitor.Listener` class like eg. 

```kotlin
class Listener {
  fun onTunnelFailureDetected(lastHandshakeEpocSeconds: Long)
}
```

And decouple whatever logic out of the monitor

File: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/configuration/WgVpnControllerService.kt [hunk 5]
```diff
@@ -126,7 +126,7 @@ interface WgVpnControllerService {
     suspend fun getEligibleLocations(): List<EligibleLocation>
 }
 
-const val NETP_ENVIRONMENT_URL = "https://controller.netp.duckduckgo.com"
+const val NETP_ENVIRONMENT_URL = "https://staging1.netp.duckduckgo.com"
```
Comment (aitorvs) [inline network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/configuration/WgVpnControllerService.kt:129]: Revert this change. Also we can set the environment from internal dev settings.

File: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/configuration/WgTunnel.kt [hunk 6]
```diff
@@ -50,15 +52,15 @@ interface WgTunnel {
...
+    suspend fun createWgConfig(keyPair: KeyPair? = null, fromFailure: Boolean = false): Result<Config>
 
     /**
      * Creates a new wireguard [Config], returns it and stores it internally.
      * If one exists, updates it and returns it and also stores it internally.
      *
      * @param keyPair is the private/public key [KeyPair] to be used in the wireguard [Config]
      */
-    suspend fun createAndSetWgConfig(keyPair: KeyPair? = null): Result<Config>
+    suspend fun createAndSetWgConfig(keyPair: KeyPair? = null, fromFailure: Boolean = false): Result<Config>
```
Comment (aitorvs) [inline network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/configuration/WgTunnel.kt:63]: Tbf, I don't like this `fromFailure` boolean value, In general I don't like flag-arguments. There's plenty of literature around that, one post I like about the topic is [this one](https://martinfowler.com/bliki/FlagArgument.html).

I have other proposals, let's discuss 1on1

**Review by aitorvs — APPROVED** [review]
Tested and works as expected
